aidial_assistant/tools_chain/tools_chain.py

```diff
@@ -0,0 +1,155 @@
+import json
+from typing import Any
+
+from aidial_sdk.chat_completion import Role
+from openai import BadRequestError
+
+from aidial_assistant.chain.callbacks.chain_callback import ChainCallback
+from aidial_assistant.chain.callbacks.command_callback import CommandCallback
+from aidial_assistant.chain.command_chain import CommandDict
+from aidial_assistant.chain.command_result import (
+    CommandInvocation,
+    CommandResult,
+    Status,
+    commands_to_text,
+    responses_to_text,
+)
+from aidial_assistant.chain.model_response_reader import (
+    AssistantProtocolException,
+)
+from aidial_assistant.commands.base import Command
+from aidial_assistant.model.model_client import (
+    ExtraResultsCallback,
+    Message,
+    ModelClient,
+    ToolCall,
+)
+from aidial_assistant.utils.open_ai import Tool
+
+
+def _publish_command(
+    command_callback: CommandCallback, name: str, arguments: str
+):
+    command_callback.on_command(name)
+    args_callback = command_callback.args_callback()
+    args_callback.on_args_start()
+    args_callback.on_args_chunk(arguments)
+    args_callback.on_args_end()
+
+
+class ToolCallsCallback(ExtraResultsCallback):
+    def __init__(self):
+        self.tool_calls: list[ToolCall] = []
+
+    def on_tool_calls(self, tool_calls: list[ToolCall]):
+        self.tool_calls = tool_calls
+
+
+class ToolsChain:
+    def __init__(
+        self,
+        model: ModelClient,
+        tools: list[Tool],
+        command_dict: CommandDict,
+    ):
+        self.model = model
+        self.tools = tools
+        self.command_dict = command_dict
+
+    async def run_chat(self, messages: list[Message], callback: ChainCallback):
+        result_callback = callback.result_callback()
+        dialogue: list[Message] = []
+        last_message_message_count = 0
+        while True:
+            tool_calls_callback = ToolCallsCallback()
+            try:
+                async for chunk in self.model.agenerate(
+                    messages + dialogue, tool_calls_callback, tools=self.tools
+                ):
+                    result_callback.on_result(chunk)
+            except BadRequestError as e:
+                if len(dialogue) == 0 or e.code == "429":
+                    raise
+
+                dialogue = dialogue[:-last_message_message_count]
+                async for chunk in self.model.agenerate(
+                    messages + dialogue, tool_calls_callback
+                ):
+                    result_callback.on_result(chunk)
+                break
+
+            if not tool_calls_callback.tool_calls:
+                break
+
+            result_messages = await self._process_tools(
```

Comment:
1. Maybe move `result_messages` one line lower, close to the only place where it's used.
2. Rename `_process_tools`  to `_run_tools`

Reply:
Moved and renamed the method to _run_tools.